packages/lodestar/src/chain/regen/queued.ts

```diff
@@ -52,72 +53,125 @@ export class QueuedStateRegenerator implements IStateRegenerator {
+      if (checkpointState) {
+        return checkpointState;
+      }
+      // The state is not immediately available in the cache, enqueue the job
+      return this.jobQueue.push({key: "getCheckpointState", args: [cp, rCaller]});
+    } catch (e) {
+      this.metrics?.regenFnTotalErrors.inc({caller, entrypoint: RegenFnName.getCheckpointState});
+      throw e;
     }
-    // The state is not immediately available in the cache, enqueue the job
-    return this.jobQueue.push({key: "getCheckpointState", args: [cp]});
   }
 
-  async getBlockSlotState(blockRoot: Root, slot: Slot): Promise<CachedBeaconState<allForks.BeaconState>> {
-    return this.jobQueue.push({key: "getBlockSlotState", args: [blockRoot, slot]});
+  async getBlockSlotState(
+    blockRoot: Root,
+    slot: Slot,
+    rCaller: IRegenCaller
+  ): Promise<CachedBeaconState<allForks.BeaconState>> {
+    const {caller} = rCaller || {};
+    this.metrics?.regenFnCallTotal.inc({caller, entrypoint: RegenFnName.getBlockSlotState});
+    return this.jobQueue.push({key: "getBlockSlotState", args: [blockRoot, slot, rCaller]});
   }
 
-  async getState(stateRoot: Root): Promise<CachedBeaconState<allForks.BeaconState>> {
-    // First attempt to fetch the state from cache before queueing
-    const state = this.stateCache.get(stateRoot);
-    if (state) {
-      return state;
+  async getState(stateRoot: Root, rCaller: IRegenCaller): Promise<CachedBeaconState<allForks.BeaconState>> {
+    const {caller} = rCaller || {};
+    this.metrics?.regenFnCallTotal.inc({caller, entrypoint: RegenFnName.getState});
+    const rmetrics = this.regen.getRegenFnMetrics({caller, entrypoint: RegenFnName.getState});
+    try {
+      // First attempt to fetch the state from cache before queueing
+      const state = this.stateCache.get(stateRoot, rmetrics);
+      if (state) {
+        return state;
+      }
+      // The state is not immediately available in the cache, enqueue the job
+      return this.jobQueue.push({key: "getState", args: [stateRoot, rCaller]});
+    } catch (e) {
+      this.metrics?.regenFnTotalErrors.inc({caller, entrypoint: RegenFnName.getPreState});
+      throw e;
     }
-    // The state is not immediately available in the cache, enqueue the job
-    return this.jobQueue.push({key: "getState", args: [stateRoot]});
   }
 
   private jobQueueProcessor = async (regenRequest: RegenRequest): Promise<CachedBeaconState<allForks.BeaconState>> => {
-    switch (regenRequest.key) {
-      case "getPreState":
-        return this.regen.getPreState(regenRequest.args[0]);
-      case "getCheckpointState":
-        return this.regen.getCheckpointState(regenRequest.args[0]);
-      case "getBlockSlotState":
-        return this.regen.getBlockSlotState(...regenRequest.args);
-      case "getState":
-        return this.regen.getState(regenRequest.args[0]);
-    }
+    const wrappedWithMetrics = async (): Promise<CachedBeaconState<allForks.BeaconState>> => {
```

Comment:
Here, why do you declare a function to then call it?

Reply:
because you are returning non awaited call in the original async jobqueueprocessor, so without looking into the queue implementation, i had to wrap it and return non awaited, and await the actual regen function inside to measure the metric.

Reply:
It doesn't matter, behavior should be identical. You can remove the wrappedWithMetrics fn